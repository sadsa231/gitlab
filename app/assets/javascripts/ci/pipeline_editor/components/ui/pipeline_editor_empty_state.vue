<script>
import { GlButton, GlSprintf } from '@gitlab/ui';
import { __ } from '~/locale';
import PipelineEditorFileNav from '~/ci/pipeline_editor/components/file_nav/pipeline_editor_file_nav.vue';

export default {
  components: {
    GlButton,
    GlSprintf,
    PipelineEditorFileNav,
  },
  i18n: {
    title: __('Optimize your workflow with CI/CD Pipelines'),
    body: __(
      'Create a new %{codeStart}.gitlab-ci.yml%{codeEnd} file at the root of the repository to get started.',
    ),
    btnText: __('Configure pipeline'),
    externalCiNote: __("This project's pipeline configuration is located outside this repository"),
    externalCiInstructions: __(
      'To edit the pipeline configuration, you must go to the project or external site that hosts the file.',
    ),
  },
  inject: {
    emptyStateIllustrationPath: {
      default: '',
    },
    usesExternalConfig: {
      default: false,
      type: Boolean,
      required: false,
    },
  },
  methods: {
    createEmptyConfigFile() {
      this.$emit('createEmptyConfigFile');
    },
  },
};
</script>
<template>
  <div>
    <pipeline-editor-file-nav v-on="$listeners" />
    <div class="gl-display-flex gl-flex-direction-column gl-align-items-center gl-mt-11">
      <img :src="emptyStateIllustrationPath" />
      <div
        v-if="usesExternalConfig"
        class="gl-display-flex gl-flex-direction-column gl-align-items-center"
      >
        <h1 class="gl-font-size-h1">{{ $options.i18n.externalCiNote }}</h1>
        <p class="gl-mt-3">{{ $options.i18n.externalCiInstructions }}</p>
      </div>
      <div v-else class="gl-display-flex gl-flex-direction-column gl-align-items-center">
        <h1 class="gl-font-size-h1">{{ $options.i18n.title }}</h1>
        <p class="gl-mt-3">
          <gl-sprintf :message="$options.i18n.body">
            <template #code="{ content }">
              <code>{{ content }}</code>
            </template>
          </gl-sprintf>
        </p>
        <gl-button
          variant="confirm"
          class="gl-mt-3"
          data-qa-selector="create_new_ci_button"
          @click="createEmptyConfigFile"
        >
          {{ $options.i18n.btnText }}
        </gl-button>
      </div>
    </div>
  </div>
</template>
