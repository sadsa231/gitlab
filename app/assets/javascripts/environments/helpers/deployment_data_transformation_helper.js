import { getIdFromGraphQLId } from '~/graphql_shared/utils';

/**
 * This function transforms Commit object coming from GraphQL to object compatible with app/assets/javascripts/vue_shared/components/commit.vue author object
 * @param {Object} Commit
 * @returns {Object}
 */
export const getAuthorFromCommit = (commit) => {
  if (commit.author) {
    return {
      username: commit.author.name,
      path: commit.author.webUrl,
      avatar_url: commit.author.avatarUrl,
    };
  }
  return {
    username: commit.authorName,
    path: `mailto:${commit.authorEmail}`,
    avatar_url: commit.authorGravatar,
  };
};

/**
 * This function transforms deploymentNode object coming from GraphQL to object compatible with app/assets/javascripts/vue_shared/components/commit.vue
 * @param {Object} deploymentNode
 * @returns {Object}
 */
export const getCommitFromDeploymentNode = (deploymentNode) => {
  if (!deploymentNode.commit) {
    throw new Error("deploymentNode argument doesn't have 'commit' field", deploymentNode);
  }
  return {
    title: deploymentNode.commit.message,
    commitUrl: deploymentNode.commit.webUrl,
    shortSha: deploymentNode.commit.shortId,
    tag: deploymentNode.tag,
    commitRef: {
      name: deploymentNode.ref,
    },
    author: getAuthorFromCommit(deploymentNode.commit),
  };
};

export const convertJobToDeploymentAction = (job) => {
  return {
    name: job.name,
    playable: job.playable,
    scheduledAt: job.scheduledAt,
    playPath: `${job.webPath}/play`,
  };
};

export const getActionsFromDeploymentNode = (deploymentNode, lastDeploymentName) => {
  if (!deploymentNode || !lastDeploymentName) {
    return [];
  }

  return (
    deploymentNode.job?.deploymentPipeline?.jobs?.nodes
      ?.filter((deployment) => deployment.name !== lastDeploymentName)
      .map(convertJobToDeploymentAction) || []
  );
};

/**
 * This function transforms deploymentNode object coming from GraphQL to object compatible with app/assets/javascripts/environments/environment_details/page.vue table
 * @param {Object} deploymentNode
 * @returns {Object}
 */
export const convertToDeploymentTableRow = (deploymentNode, environment) => {
  const { lastDeployment } = environment;
  const commit = getCommitFromDeploymentNode(deploymentNode);
  return {
    status: deploymentNode.status.toLowerCase(),
    id: deploymentNode.iid,
    triggerer: deploymentNode.triggerer,
    commit,
    job: deploymentNode.job && {
      webPath: deploymentNode.job.webPath,
      label: `${deploymentNode.job.name} (#${getIdFromGraphQLId(deploymentNode.job.id)})`,
    },
    created: deploymentNode.createdAt || '',
    deployed: deploymentNode.finishedAt || '',
    actions: getActionsFromDeploymentNode(deploymentNode, lastDeployment?.job?.name),
  };
};
