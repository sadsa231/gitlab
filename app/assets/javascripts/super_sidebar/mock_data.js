import { s__ } from '~/locale';

export const context = {
  title: 'Typeahead.js',
  link: '/',
  avatar: 'https://gitlab.com/uploads/-/system/project/avatar/278964/project_avatar.png?width=32',
};

export const contextSwitcherItems = {
  yourWork: { title: s__('Navigation|Your work'), link: '/', icon: 'work' },
  recentProjects: [
    {
      // eslint-disable-next-line @gitlab/require-i18n-strings
      title: 'Orange',
      subtitle: 'tropical-tree',
      link: '/tropical-tree',
      avatar:
        'https://gitlab.com/uploads/-/system/project/avatar/4456656/pajamas-logo.png?width=64',
    },
    {
      // eslint-disable-next-line @gitlab/require-i18n-strings
      title: 'Lemon',
      subtitle: 'tropical-tree',
      link: '/tropical-tree',
      avatar: 'https://gitlab.com/uploads/-/system/project/avatar/7071551/GitLab_UI.png?width=64',
    },
    {
      // eslint-disable-next-line @gitlab/require-i18n-strings
      title: 'Coconut',
      subtitle: 'tropical-tree',
      link: '/tropical-tree',
      avatar:
        'https://gitlab.com/uploads/-/system/project/avatar/4149988/SVGs_project.png?width=64',
    },
  ],
  recentGroups: [
    {
      title: 'Developer Evangelism at GitLab',
      subtitle: 'tropical-tree',
      link: '/tropical-tree',
      avatar:
        'https://gitlab.com/uploads/-/system/group/avatar/10087220/rainbow_tanuki.jpg?width=64',
    },
    {
      title: 'security-products',
      subtitle: 'tropical-tree',
      link: '/tropical-tree',
      avatar:
        'https://gitlab.com/uploads/-/system/group/avatar/11932235/gitlab-icon-rgb.png?width=64',
    },
    {
      title: 'Tanuki-Workshops',
      subtitle: 'tropical-tree',
      link: '/tropical-tree',
      avatar:
        'https://gitlab.com/uploads/-/system/group/avatar/5085244/Screenshot_2019-04-29_at_16.13.07.png?width=64',
    },
  ],
};
