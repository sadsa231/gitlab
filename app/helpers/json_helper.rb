# frozen_string_literal: true

module JsonHelper
  # These two JSON helpers are short-form wrappers for the Gitlab::Json
  # class, which should be used in place of .to_json calls or calls to
  # the JSON class.
  def json_generate(*args)
    Gitlab::Json.generate(*args)
  end

  def json_parse(*args)
    Gitlab::Json.parse(*args)
  end
end
