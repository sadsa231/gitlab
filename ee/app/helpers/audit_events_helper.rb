# frozen_string_literal: true

module AuditEventsHelper
  FILTER_TOKEN_TYPES = {
    user: :user,
    group: :group,
    project: :project,
    member: :member
  }.freeze

  def admin_audit_event_tokens
    [
      { type: FILTER_TOKEN_TYPES[:user] },
      { type: FILTER_TOKEN_TYPES[:group] },
      { type: FILTER_TOKEN_TYPES[:project] }
    ].freeze
  end

  def group_audit_event_tokens(group_id)
    [{ type: FILTER_TOKEN_TYPES[:member], group_id: group_id }].freeze
  end

  def project_audit_event_tokens(project_path)
    [{ type: FILTER_TOKEN_TYPES[:member], project_path: project_path }].freeze
  end

  def export_url
    admin_audit_log_reports_url(format: :csv)
  end

  def view_only_own_project_events?(project)
    !can?(current_user, :admin_project, project) && !current_user.auditor?
  end

  def view_only_own_group_events?(group)
    !can?(current_user, :admin_group, group) && !current_user.auditor?
  end

  def filter_view_only_own_events_token_values(view_only)
    return [] unless view_only

    [{ type: FILTER_TOKEN_TYPES[:member], data: "@#{current_user.username}" }]
  end

  def show_streams_for_group?(group)
    return false if group.subgroup?

    can?(current_user, :admin_external_audit_events, group)
  end
end
