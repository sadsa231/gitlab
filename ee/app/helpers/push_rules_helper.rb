# frozen_string_literal: true

module PushRulesHelper
  def can_modify_group_push_rules?(current_user, group)
    can?(current_user, :change_push_rules, group)
  end

  def reject_unsigned_commits_description(push_rule)
    message = s_("ProjectSettings|Only signed commits can be pushed to this repository.")

    push_rule_update_description(message, push_rule, :reject_unsigned_commits)
  end

  def reject_non_dco_commits_description(push_rule)
    message = format(s_("ProjectSettings|Only commits that include a %{code_block_start}Signed-off-by:%{code_block_end} element can be pushed to this repository.").html_safe, code_block_start: '<code>'.html_safe, code_block_end: '</code>'.html_safe)

    push_rule_update_description(message, push_rule, :reject_non_dco_commits)
  end

  def commit_committer_check_description(push_rule)
    message = s_("ProjectSettings|Users can only push commits to this repository "\
                 "if the committer email is one of their own verified emails.")

    push_rule_update_description(message, push_rule, :commit_committer_check)
  end

  def commit_committer_name_check_description(push_rule)
    message = s_("ProjectSettings|Users can only push commits to this repository "\
      "if the commit author name is consistent with their GitLab account name.")

    push_rule_update_description(message, push_rule, :commit_committer_name_check)
  end

  private

  def push_rule_update_description(message, push_rule, rule)
    messages = [message]
    if push_rule.global?
      messages << s_("ProjectSettings|This setting will be applied to all projects unless overridden by an admin.")
    else
      enabled_globally = PushRule.global&.public_send(rule)
      enabled_in_project = push_rule.public_send(rule) # rubocop:disable GitlabSecurity/PublicSend

      if enabled_globally
        messages << if enabled_in_project
                      s_("ProjectSettings|This setting is applied on the server level and can be overridden by an admin.")
                    else
                      s_("ProjectSettings|This setting is applied on the server level but has been overridden for this project.")
                    end

        unless current_user.can_admin_all_resources?
          messages << s_("ProjectSettings|Contact an admin to change this setting.")
        end
      end
    end

    messages.join(' ').html_safe
  end
end
