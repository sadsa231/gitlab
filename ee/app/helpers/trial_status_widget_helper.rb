# frozen_string_literal: true

# NOTE: The patterns first introduced in this helper for doing trial-related
# callouts are mimicked by the PaidFeatureCalloutHelper. A third reuse of these
# patterns (especially as these experiments finish & become permanent parts of
# the codebase) could trigger the need to extract these patterns into a single,
# reusable, sharable helper.
module TrialStatusWidgetHelper
  def trial_status_popover_data_attrs(group, trial_status, ultimate_plan_id)
    hand_raise_attrs =
      experiment(
        :group_contact_sales, namespace: group.root_ancestor, user: current_user, sticky_to: current_user
      ) do |e|
        e.control { {} }
        e.candidate { hand_raise_props(group, glm_content: 'trial-status-show-group') }
      end.run

    base_attrs = trial_status_common_data_attrs(group).merge(hand_raise_attrs)

    base_attrs.merge(
      days_remaining: trial_status.days_remaining, # for experiment tracking
      group_name: group.name,
      purchase_href: ultimate_subscription_path_for_group(group, ultimate_plan_id),
      target_id: base_attrs[:container_id],
      trial_end_date: trial_status.ends_on
    )
  end

  def trial_status_widget_data_attrs(group, trial_status)
    trial_status_common_data_attrs(group).merge(
      trial_days_used: trial_status.days_used,
      trial_duration: trial_status.duration,
      nav_icon_image_path: image_path('illustrations/golden_tanuki.svg'),
      percentage_complete: trial_status.percentage_complete
    )
  end

  private

  def trial_status_common_data_attrs(group)
    {
      container_id: 'trial-status-sidebar-widget',
      plan_name: group.gitlab_subscription.plan_title,
      plans_href: group_billings_path(group)
    }
  end

  def ultimate_subscription_path_for_group(group, ultimate_plan_id)
    new_subscriptions_path(namespace_id: group.id, plan_id: ultimate_plan_id)
  end
end

TrialStatusWidgetHelper.prepend_mod
