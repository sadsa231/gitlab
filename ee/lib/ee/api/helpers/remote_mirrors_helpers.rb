# frozen_string_literal: true

module EE
  module API
    module Helpers
      module RemoteMirrorsHelpers
        extend ActiveSupport::Concern
        extend ::Gitlab::Utils::Override

        prepended do
          params :mirror_branches_setting_ee do
            optional :mirror_branch_regex, type: String, desc: 'Determines if only matched branches are mirrored'
            mutually_exclusive :only_protected_branches, :mirror_branch_regex
          end
        end

        override :verify_mirror_branches_setting
        def verify_mirror_branches_setting(attrs, project)
          attrs[:mirror_branch_regex] = nil unless ::Feature.enabled?(:mirror_only_branches_match_regex, project)

          if attrs[:only_protected_branches]
            attrs[:mirror_branch_regex] = nil
          elsif attrs[:mirror_branch_regex].present?
            attrs[:only_protected_branches] = false
          end
        end
      end
    end
  end
end
