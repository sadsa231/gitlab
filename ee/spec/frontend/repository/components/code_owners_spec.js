import { GlLink, GlButton } from '@gitlab/ui';
import { shallowMount, mount } from '@vue/test-utils';
import Vue, { nextTick } from 'vue';
import VueApollo from 'vue-apollo';
import createMockApollo from 'helpers/mock_apollo_helper';
import waitForPromises from 'helpers/wait_for_promises';
import CodeOwners from 'ee_component/vue_shared/components/code_owners/code_owners.vue';
import codeOwnersInfoQuery from 'ee/graphql_shared/queries/code_owners_info.query.graphql';
import { extendedWrapper } from 'helpers/vue_test_utils_helper';
import { codeOwnerMock, codeOwnersMultipleMock, codeOwnersDataMock, refMock } from '../mock_data';

let wrapper;
let mockResolver;

const createComponent = async (codeOwners = [codeOwnerMock], mountFn = shallowMount) => {
  Vue.use(VueApollo);

  const project = {
    ...codeOwnersDataMock,
    repository: {
      blobs: {
        nodes: [{ id: '345', codeOwners }],
      },
    },
  };

  mockResolver = jest.fn().mockResolvedValue({ data: { project } });

  wrapper = extendedWrapper(
    mountFn(CodeOwners, {
      apolloProvider: createMockApollo([[codeOwnersInfoQuery, mockResolver]]),
      propsData: { projectPath: 'some/project', filePath: 'some/file' },
      mixins: [{ data: () => ({ ref: refMock }) }],
    }),
  );

  await waitForPromises();
};

describe('Code owners component', () => {
  const findHelpIcon = () => wrapper.findByTestId('help-icon');
  const findUsersIcon = () => wrapper.findByTestId('users-icon');
  const findCodeOwners = () => wrapper.findAllByTestId('code-owners');
  const findCommaSeparators = () => wrapper.findAllByTestId('comma-separator');
  const findAndSeparators = () => wrapper.findAllByTestId('and-separator');
  const findToggle = () => wrapper.findComponent(GlButton);
  const findLink = () => wrapper.findComponent(GlLink);

  beforeEach(() => createComponent());

  afterEach(() => wrapper.destroy());

  describe('help link', () => {
    it('renders a GlLink component', () => {
      expect(findLink().exists()).toBe(true);
      expect(findLink().attributes('href')).toBe('/help/user/project/code_owners');
      expect(findLink().attributes('target')).toBe('_blank');
      expect(findLink().attributes('title')).toBe('About this feature');
    });

    it('renders a Help icon', () => {
      expect(findHelpIcon().exists()).toBe(true);
      expect(findHelpIcon().props('name')).toBe('question-o');
    });
  });

  it('renders a Users icon', () => {
    expect(findUsersIcon().exists()).toBe(true);
    expect(findUsersIcon().props('name')).toBe('users');
  });

  it('doesn`t render toggle when the number of codeowners is less than 5', () => {
    expect(findToggle().exists()).toBe(false);
  });

  it.each`
    codeOwners                            | commaSeparators | codeOwnersLength
    ${[]}                                 | ${0}            | ${0}
    ${[codeOwnerMock]}                    | ${0}            | ${1}
    ${codeOwnersMultipleMock.slice(0, 2)} | ${1}            | ${2}
    ${codeOwnersMultipleMock.slice(0, 3)} | ${2}            | ${3}
    ${codeOwnersMultipleMock}             | ${4}            | ${5}
  `('matches the snapshot', async ({ codeOwners, commaSeparators, codeOwnersLength }) => {
    await createComponent(codeOwners);

    expect(findCommaSeparators().length).toBe(commaSeparators);
    expect(findCodeOwners().length).toBe(codeOwnersLength);
    expect(wrapper.element).toMatchSnapshot();
  });

  describe('when the number of code owners is more than 5', () => {
    beforeEach(() => createComponent(codeOwnersMultipleMock, mount));

    it('renders a toggle button with correct text', () => {
      expect(findToggle().exists()).toBe(true);
      expect(findToggle().text()).toBe('2 more');
    });

    it('renders only first 5 codeowners at initial state', () => {
      expect(findCodeOwners().length).toBe(5);
    });

    it('expands when you click on a toggle', async () => {
      await findToggle().trigger('click');
      await nextTick();
      expect(findCodeOwners().length).toBe(codeOwnersMultipleMock.length);
      expect(findToggle().text()).toBe('show less');
      expect(findAndSeparators().length).toBe(1);
    });
  });
});
