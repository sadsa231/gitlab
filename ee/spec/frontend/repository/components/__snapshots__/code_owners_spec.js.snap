// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Code owners component matches the snapshot 1`] = `<!---->`;

exports[`Code owners component matches the snapshot 2`] = `
<div
  class="well-segment blob-auxiliary-viewer file-owner-content gl-display-flex gl-flex-wrap"
>
  <div
    class="gl-display-inline gl-mr-2"
  >
    <gl-icon-stub
      data-testid="users-icon"
      name="users"
      size="16"
    />
     
    <strong>
      Code owners
    </strong>
     
    <gl-link-stub
      href="/help/user/project/code_owners"
      target="_blank"
      title="About this feature"
    >
      <gl-icon-stub
        data-testid="help-icon"
        name="question-o"
        size="16"
      />
    </gl-link-stub>
    
    :
  
  </div>
   
  <div
    data-testid="code-owners"
  >
    <!---->
     
    <gl-link-stub
      href="path/to/@johnDoe"
      target="_blank"
    >
      
      John Doe
    
    </gl-link-stub>
  </div>
   
  <!---->
   
  <!---->
</div>
`;

exports[`Code owners component matches the snapshot 3`] = `
<div
  class="well-segment blob-auxiliary-viewer file-owner-content gl-display-flex gl-flex-wrap"
>
  <div
    class="gl-display-inline gl-mr-2"
  >
    <gl-icon-stub
      data-testid="users-icon"
      name="users"
      size="16"
    />
     
    <strong>
      Code owners
    </strong>
     
    <gl-link-stub
      href="/help/user/project/code_owners"
      target="_blank"
      title="About this feature"
    >
      <gl-icon-stub
        data-testid="help-icon"
        name="question-o"
        size="16"
      />
    </gl-link-stub>
    
    :
  
  </div>
   
  <div
    data-testid="code-owners"
  >
    <!---->
     
    <gl-link-stub
      href="/raisa"
      target="_blank"
    >
      
      Idella Welch
    
    </gl-link-stub>
  </div>
  <div
    data-testid="code-owners"
  >
    <span
      data-testid="comma-separator"
    >
      , 
    </span>
     
    <gl-link-stub
      href="/noella"
      target="_blank"
    >
      
      Winston Von
    
    </gl-link-stub>
  </div>
   
  <!---->
   
  <!---->
</div>
`;

exports[`Code owners component matches the snapshot 4`] = `
<div
  class="well-segment blob-auxiliary-viewer file-owner-content gl-display-flex gl-flex-wrap"
>
  <div
    class="gl-display-inline gl-mr-2"
  >
    <gl-icon-stub
      data-testid="users-icon"
      name="users"
      size="16"
    />
     
    <strong>
      Code owners
    </strong>
     
    <gl-link-stub
      href="/help/user/project/code_owners"
      target="_blank"
      title="About this feature"
    >
      <gl-icon-stub
        data-testid="help-icon"
        name="question-o"
        size="16"
      />
    </gl-link-stub>
    
    :
  
  </div>
   
  <div
    data-testid="code-owners"
  >
    <!---->
     
    <gl-link-stub
      href="/raisa"
      target="_blank"
    >
      
      Idella Welch
    
    </gl-link-stub>
  </div>
  <div
    data-testid="code-owners"
  >
    <span
      data-testid="comma-separator"
    >
      , 
    </span>
     
    <gl-link-stub
      href="/noella"
      target="_blank"
    >
      
      Winston Von
    
    </gl-link-stub>
  </div>
  <div
    data-testid="code-owners"
  >
    <span
      data-testid="comma-separator"
    >
      , 
    </span>
     
    <gl-link-stub
      href="/edyth"
      target="_blank"
    >
      
      Don Runte
    
    </gl-link-stub>
  </div>
   
  <!---->
   
  <!---->
</div>
`;

exports[`Code owners component matches the snapshot 5`] = `
<div
  class="well-segment blob-auxiliary-viewer file-owner-content gl-display-flex gl-flex-wrap"
>
  <div
    class="gl-display-inline gl-mr-2"
  >
    <gl-icon-stub
      data-testid="users-icon"
      name="users"
      size="16"
    />
     
    <strong>
      Code owners
    </strong>
     
    <gl-link-stub
      href="/help/user/project/code_owners"
      target="_blank"
      title="About this feature"
    >
      <gl-icon-stub
        data-testid="help-icon"
        name="question-o"
        size="16"
      />
    </gl-link-stub>
    
    :
  
  </div>
   
  <div
    data-testid="code-owners"
  >
    <!---->
     
    <gl-link-stub
      href="/raisa"
      target="_blank"
    >
      
      Idella Welch
    
    </gl-link-stub>
  </div>
  <div
    data-testid="code-owners"
  >
    <span
      data-testid="comma-separator"
    >
      , 
    </span>
     
    <gl-link-stub
      href="/noella"
      target="_blank"
    >
      
      Winston Von
    
    </gl-link-stub>
  </div>
  <div
    data-testid="code-owners"
  >
    <span
      data-testid="comma-separator"
    >
      , 
    </span>
     
    <gl-link-stub
      href="/edyth"
      target="_blank"
    >
      
      Don Runte
    
    </gl-link-stub>
  </div>
  <div
    data-testid="code-owners"
  >
    <span
      data-testid="comma-separator"
    >
      , 
    </span>
     
    <gl-link-stub
      href="/julietta_rogahn"
      target="_blank"
    >
      
      Sherri McClure
    
    </gl-link-stub>
  </div>
  <div
    data-testid="code-owners"
  >
    <span
      data-testid="comma-separator"
    >
      , 
    </span>
     
    <gl-link-stub
      href="/youlanda"
      target="_blank"
    >
      
      Amber Koch
    
    </gl-link-stub>
  </div>
   
  <!---->
   
  <span
    class="gl-ml-2"
  >
     and 
  </span>
   
  <gl-button-stub
    buttontextclasses=""
    category="primary"
    class="gl-vertical-align-text-bottom gl-ml-2"
    icon=""
    size="medium"
    variant="link"
  >
    
      2 more
    
  </gl-button-stub>
</div>
`;
